### ISSUE TITLE
Is Chrome launched differently through Canopy than when I launch from my desktop?

### ISSUE THREAD

**Issue by sprice9161** [lefthandedgoat/canopy#325]
I have a chrome extension that I use called selectorgadget that I use to identify CSS selectors on my web pages. 

When I launch chrome using my desktop icon (which points to the chrome installation C:\Program Files (x86)\Google\Chrome\Application\chrome.exe ), chrome launches and the previously loaded extension appears in the toolbar every time. Also, selecting the menu icon, and selecting "More Tools" displays "Extensions" as enabled.

When I execute my automation script, with a breakpoint set so that I can view the browser in a static state, there is NO icon displayed in the toolbar, and the menu icon, and selecting "More Tools" displays "Extensions" as Disabled. 

The latter scenario is clearly not what I want since I require this extension for debugging purposes. 
.......
Okay just looked into what chromedriver.exe is. I didn't realize that this is essentially for launching chrome specifically for automation, replacing chrome.exe. I located a web page with tons of options but I have no idea which one to use. Do you have experience with them? Running chromedriver.exe --help only displays 9 of the listed option, which is confusing.

**Comment by sprice9161** [comment]
...Does the Canopy configuration statement allow me to pass one of the command line parameters?

**Comment by lefthandedgoat** [comment]
Chromedriver does not replace chrome.  It launches chrome for you and you talk to chrome through chromedriver.

Selenium intentionally launches chrome without plugins enabled because plugins can do mostly arbitrary things and break the site you are working on, which would impede your ability to test it.

If you want to launch chrome with extensions you need to download the extension and manually load it with chrome options (never tried, just read a stack overflow post about it). http://stackoverflow.com/questions/34222412/load-chrome-extension-using-selenium

**Comment by lefthandedgoat** [comment]
For your second question, you probably want to use this https://github.com/lefthandedgoat/canopy/blob/master/src/canopy/canopy.fs#L809

And if you look at the lines above it looks like the basic launching of chrome calls --disable-options

**Comment by lefthandedgoat** [comment]
Also your tool uses bookmarklets and I have 0 clue how that plays into this.

You can right click an element and choose `Inspect` from the context menu, then use the dev console to test your selectors with `$(".myClass")`

Learning write your own selectors well is what makes UI Automation from OK to really useful.

**Comment by sprice9161** [comment]
Saw that stackflow yesterday. it requires a .crx file, and I can not locate one in the zip that I downloaded. 

Can place that line of code that you referenced after the chromeDir statement? Adding an option of course.

So here is the thing about writing the selectors. This tool is soooooooo much easier, and it allows me to review all of the selectors on the page, where they should already have identifiers provided by the developers.

**Comment by lefthandedgoat** [comment]
Other answers suggest you get the crx with this:

https://chrome.google.com/webstore/detail/get-crx/dijpllakibenlejkbajahncialkbdkjc


```
let options = Chrome.ChromeOptions()
options.AddArguments("whatever options you want")
options.AddArguments("more options")

start (ChromeWithOptions options)         
```

**Comment by lefthandedgoat** [comment]
Here is a guide to use the built in chrome dev tools which will also let you review your selectors on the page.

http://yizeng.me/2014/03/23/evaluate-and-validate-xpath-css-selectors-in-chrome-developer-tools/

**Comment by sprice9161** [comment]
Thank you Chris. I left on Friday, and saw your emails on my way home. I'll get back to you.

**Comment by sprice9161** [comment]
Hello Chris,
I have successfully attained the .crx file. It failed to install. This is an internal issue with my access rights that I am working through with our help desk.

IF you want to close this ticket it's okay. I won't be able to attempt the programmatic example for launching from VS until I've resolved my access rights.

Thank you for the additional info that you provided me. I've bookmarked all of it.

Regards

Steve
